packages/fluentui/react-builder/src/narration/NarrationComputer.tsx

```diff
@@ -23,14 +35,186 @@ export class NarrationComputer {
+    element: IAriaElement,
+    path: string[],
+    focusableElements: FocusableElement[],
+    activeDescendantsParents: IAriaElement[],
+  ) {
+    // Determine the path
+    const newPath = path.slice();
+    const node = await (window as any).getComputedAccessibleNode(element);
+
+    // Only add the item to the path if the role is not a generic container
+    if (node.role !== 'genericContainer') {
+      const item = node.name ? `${node.role} (${node.name})` : node.role;
+      newPath.push(item);
+    }
+
+    // If the element is focusable, save it together with its path
+    const isDirectlyFocusable =
+      (element.getAttribute('tabindex') || element.tabIndex >= 0) && element.ariaActiveDescendantElement == null;
+    const isActiveDescendant = activeDescendantsParents.some(parent => {
+      // Begin some 1
+      return parent.tabIndex >= 0 && parent.ariaActiveDescendantElement === element;
+    }); // End some 1
+    if (isDirectlyFocusable || isActiveDescendant) {
+      // Begin if 1
+      const focusableElement: FocusableElement = {
+        path: newPath,
+        element,
+      };
+      focusableElements.push(focusableElement);
+      // return;
+    } // End if 1
+
+    // Traverse down to all the children
+    for (let i = 0; i < element.children.length; i++) {
+      // Begin for 1
+      const child = element.children[i] as IAriaElement;
+      await this.findFocusableElements(child, newPath, focusableElements, activeDescendantsParents);
+    } // End for 1
+  } // End findFocusableElements
+
+  // Returns ARIA landmarks and groups that would be entered and narrated when focus moves from the given previous element to the given element, for the given platform.
+  getEnteredLandmarksAndGroups(element: HTMLElement, prevElement: HTMLElement, platform: string): string[] {
+    const landmarksAndGroups = [];
+    let elementParent = element.parentElement;
+    let prevElementParent = prevElement.parentElement;
+    let commonAncestor = null;
+
+    // Find the common ancestor of the previous and current element
+    while (elementParent !== prevElementParent && elementParent !== element.ownerDocument.body) {
+      // Begin while 1
+      while (elementParent !== prevElementParent && prevElementParent !== prevElement.ownerDocument.body) {
+        // Begin while 2
+        // If the element is an ancestor of the previous element, return
+        if (prevElementParent === element) {
+          // Begin if 1
+          return landmarksAndGroups;
+        } // End if 1
+        prevElementParent = prevElementParent.parentElement;
+      } // End while 2
+      // If the previous element is an ancestor of the element, return
+      if (elementParent === prevElement) {
+        // Begin if 1
+        return landmarksAndGroups;
+      } // End if 1
+      prevElementParent = prevElement.parentElement;
+      elementParent = elementParent.parentElement;
+    } // End while 1
+    commonAncestor = elementParent;
```

Comment:
Maybe this could be abstracted to be a function, at I think it's doing here is traversing up the tree to return the common ancestor... maybe a function called `getCommonAncestor(elementParent, prevElementParent, element)` ?

It will also make this code easier to read

Reply:
Agree, I have extracted the function.